### DISCUSSION TITLE
Start At Time, Detect Motion, Stop At Time So Auto Upload Can Run?

### DISCUSSION THREAD

**Discussion by cstan22** [gopro/labs#1050] [General]
Anyone know how to make something like this work? I'd think something like !13:30N!SM5-5H3!R would start the camera at 13:30 and then detect for motion on repeat, but it doesn't work. It only starts the camera, but doesn't look for motion. And when trying to add "end after" in the Labs app, it doesn't add any characters to the QR command if motion is being used. Appreciate the help!

**Comment by dnewman-gpsw** [comment]
You can't use the old !SM command with a stop time, it was designed with that in mind. You have to use the new scripting language and that is a stepper learning curve.  And scripting uploads is also tricky. 

Use will need to use the motion variable.

From docs: https://gopro.github.io/labs/control/tech/
* m motionValue - <mValueCMD if(motion < Value) CMD Motion value is a percentage of pixels moving e.g. >m5!S+60E!R, this look of motion greater than 5%, and record for 60 seconds when detected.

Also also will likely benefit from the visual script builder: https://gopro.github.io/labs/build/

**Reply by cstan22** [reply]
Hi - appreciate you taking the time to respond. Ignoring the upload feature, trying >m5!S+60E!R does nothing. GoPro Hero9 scans the QR code, but nothing happens, and then I can't scan any new QR codes without restarting the device. I also tried generating a custom code using the visual script builder. Got this: =Am>A5!S==A0!E. Same issue.

Any guidance on how to get it to work? And how to add the upload feature? Or where to look for a more comprehensive guide? Uploading doesn't have to be after each video record. Could be at the end of the day or something. Mostly, I want to be able capture motion activated events throughout the day, automatically upload them by the end of the day, and then restart the process automatically. Thanks!

**Reply by cstan22** [reply]
FWIW, something like this kind of works: !SM3D4H5!U!R

Problem is that after the upload, the GoPro doesn't advance to restart script until the screen is touched. I tried introducing an extra step after !U, but the only thing that seems to work is the turn off and restart command, but then the script is cancelled. Any thoughts?

**Comment by dnewman-gpsw** [comment]
Sorry, I assumed you had a more recent camera. All the advice I gave is only for HERO10 onward.  HERO9 only has the old style motion detection and you have already come across its limitations.  It works fine if you're using it as designed, but you can't schedule it to stop motion detection at a certain time.

**Reply by cstan22** [reply]
I see. I was trying to test on my old Hero9 before ordering some new ones. On the newer models, does it avoid the issue I describe above where the script gets stuck on the "Videos Uploaded" screen until I touch the screen?

Also, is there a way using the new scripting language to have it run motion detection throughout the day, and then at a certain time, run upload, and then restart the script? Thanks!

**Reply by dnewman-gpsw** [reply]
Yes, it possible, although the main issue is upload.  On a reliable network it should work, but the infield applications the network connection is always the weak link. Labs is not guaranteed to have any feature, include motion detection, as Labs is for experimenting and fun. So you shouldn't make a purchase decision based on Labs features, but Labs is better on later models.  And Labs is coming for HERO13.

**Reply by cstan22** [reply]
Thanks. Would be on a reliable network. And understood regarding Labs. Any ideas on how to schedule the upload along with the motion detection? I feel like adding something like !19:00U before !R in >m5!S+60E!R would cause it to motion detect once and then wait until 19:00 to do anything else

**Reply by dnewman-gpsw** [reply]
You just have the code up the rules. Work out the logic and the code using the visual script builder: https://gopro.github.io/labs/build/ once you have a newer camera.

**Reply by cstan22** [reply]
<img width="566" alt="2024-09-05_16-57-02" src="https://github.com/user-attachments/assets/3d057052-b39d-4bb7-97ad-8e350658a916">

Do you think something like this would work? And if so, do you know how I could add a "hold" to the end capture command?

**Reply by dnewman-gpsw** [reply]
That particular script will not work, but you are getting the right idea. You can't put if conditions within the else statement.  Just remove all the elements from the else and just put it at the end of the loop, it will work fine.  Yes, you can program the hold tool.

**Reply by cstan22** [reply]
Ok, how about this. In my mind, this says: If the current time is between 21:00 and 21:04, then upload to cloud. Next, if current time is before 21:00, record the current level of motion to a variable and check to see if the camera is NOT recording. If it isn't, check to see if motion is at least 5%. If so, begin capture. Then check to see if the camera IS recording. If so, set a new variable (B) to 0 and begin a loop. On each loop, check to see if motion is at 0%. If it is, add 1 to variable B and then check to see if variable B is less than 5. If it is, restart the loop. As soon as variable B is >= 5, end the motion capture. Then, do the same script where current time is before 21:00 for a situation where current time is above 21:04. Then repeat the whole script.

So, the script always tries to upload to cloud first but is only successful between 21:00 and 21:04 and otherwise is always repeating a motion activation camera capture with logic to end the capture on a hold of 5.

Think that would work?

One thing that's unclear to me: how long does a loop take? Like, each time variable B is added to, how much time passes? Ie, would counting variable B up to 5, take 5 seconds? If not, how can I do that better?

Thanks!

<img width="325" alt="2024-09-05_17-31-16" src="https://github.com/user-attachments/assets/feb44bce-a3fd-4a90-9b8f-bb0c1bc42a76">

**Reply by dnewman-gpsw** [reply]
Sorry, that will not work either.  As the scripting language doesn't use braces, you have to more particular with IF statements.

if(if(if(thing1)else(thing2)) is okay
but if(thing1 if(thing2)) is not <- you have a lot of this structure

You can learn of this once you are experiment with a Labs advanced scripting enabled camera.

**Reply by cstan22** [reply]
Are you saying that I can't have two IF statements inside of a then? And if so, does replacing them with an if-else-then like this image fix it? Or that I can't do something like if current time is, then set variable and then if camera not recording, do something? If that's not possible, how would I set a variable before doing an if based on that variable like in the loop that I have? Thanks!

<img width="318" alt="2024-09-05_19-58-02" src="https://github.com/user-attachments/assets/df722a3d-da08-497b-b384-df84bad536e9">

**Reply by cstan22** [reply]
I found this "status loop count" option, so I think I was able to update it per your guidance (with only an action or an if within a then instead of an action followed by an if within a then). Would be appreciative for any feedback. Either way, will test on Sat when the new GoPro arrives. 

Is it possible to do like an OR if statement so I can have just one set that says if current time is less than 19:00 OR greater than 19:04, so that I don't have to repeat the whole script for each if condition? Thanks!

<img width="374" alt="2024-09-06_01-12-50" src="https://github.com/user-attachments/assets/0ddc1e38-8d47-4ad2-916c-5efb96954704">

**Reply by dnewman-gpsw** [reply]
There is no OR.  Everything can all be done with IFs only. Design that why and optimize later.

You are still ignoring the warning "Only use on the inner-most of nested ifs" of the IF-THEN.

Else should only contain settings or action, not more conditions.

**Reply by cstan22** [reply]
Ok, feeling good about this one. Used your guidance on reversing ORs, and I don't think I'm breaking any of the IF nesting rules

<img width="517" alt="2024-09-06_13-10-04" src="https://github.com/user-attachments/assets/96a8650c-de8f-45b2-820b-e594095a8e23">

**Reply by dnewman-gpsw** [reply]
Loop: is not in an expect position, it might work, untested.  Loop count is everything, so that will work.

This is how you could implement the hold. Here is A to 1 when I want to start counting.
![image](https://github.com/user-attachments/assets/04392212-e32b-4eb8-89e6-52ab12df6f2e)

**Reply by cstan22** [reply]
Are you saying Loop isn't allowed to be in a then? Sorry I don't understand. And what does Sleep do? It pauses the script for X second(s)? Thank you!

**Reply by cstan22** [reply]
Ok, I got a GoPro 11 and have tried a bunch of different builds, but the problem seems to be that I have to touch the screen after end capture to get it to work again. Any way to fix this?

**Comment by cstan22** [comment]
I'm hitting a wall. Ended up with something like this. After scanning the QR code, it records immediately, and after it ends capture, it gets buggy. Sometimes it recognizes motion. Sometimes not. When it does, it's often pretty delayed in starting / ending captures. The script will cycle through printing "Recycle" and "No motion", leading me to think that there's some bug that stops it from activating motion detection again after each round.

Any thoughts are appreciated!

<img width="424" alt="2024-09-09_11-55-57" src="https://github.com/user-attachments/assets/da8ac0f0-d84d-49f2-84aa-19c83f3eb6d7">

**Reply by dnewman-gpsw** [reply]
Experiment with on the motion detector, not the hold, not the upload. Get that right first.  You can also print variables.  $A will print variable A.
